spinalcordtoolbox/centerline/core.py

```diff
@@ -173,11 +175,28 @@ def get_centerline(im_seg, param=ParamCenterline(), verbose=1, remove_temp_files
         # Create an image with the centerline
         im_centerline = im_seg.copy()
         im_centerline.data = np.zeros(im_centerline.data.shape)
-        # Assign value=1 to centerline. Make sure to clip to avoid array overflow.
-        # TODO: check this round and clip-- suspicious
-        im_centerline.data[round_and_clip(x_centerline_fit, clip=[0, im_centerline.data.shape[0]]),
-                           round_and_clip(y_centerline_fit, clip=[0, im_centerline.data.shape[1]]),
-                           z_ref] = 1
+        # Binarized
+        if param.soft == 0:
+            # Assign value=1 to centerline. Make sure to clip to avoid array overflow.
+            im_centerline.data[round_and_clip(x_centerline_fit, clip=[0, im_centerline.data.shape[0]]),
+                               round_and_clip(y_centerline_fit, clip=[0, im_centerline.data.shape[1]]),
+                               z_ref] = 1
+        # Soft (accounting for partial volume effect)
+        else:
+            # Compute integer pixel locations (all together, these form a 2x2 grid around centerline)
+            x_ceil = np.ceil(x_centerline_fit).astype(int)
+            y_ceil = np.ceil(y_centerline_fit).astype(int)
+            x_floor = np.floor(x_centerline_fit).astype(int)
+            y_floor = np.floor(y_centerline_fit).astype(int),
+            # Compute the difference between the centerline and its rounded values
+            x_diff = x_centerline_fit - x_centerline_fit.astype(int)
+            y_diff = y_centerline_fit - y_centerline_fit.astype(int)
+            # Assign soft segmentation values to the 2x2 grid based on the computed differences
+            im_centerline.data[x_ceil, y_ceil, z_ref] = 0.5 * (x_diff + y_diff)
+            im_centerline.data[x_ceil, y_floor, z_ref] = 0.5 * (x_diff + (1 - y_diff))
+            im_centerline.data[x_floor, y_ceil, z_ref] = 0.5 * ((1 - x_diff) + y_diff)
+            im_centerline.data[x_floor, y_floor, z_ref] = 0.5 * ((1 - x_diff) + (1 - y_diff))
```

Comment:
Also, I just want to sanity check that the distribution of `x/y_diff` and `(1 - x/y_diff)` is correct:

Say, for example, that `x_centerline_fit` and `y_centerline_fit` are both at voxel location `10.75`:

- `ceil` == `11`
- `floor` == `10`
- `diff` == `0.75`
- `[11,11]` = `0.25 * (0.75 + 0.75)` = `0.375`
- `[11,10]` = `0.25 * (0.75 + 0.25)` = `0.25`
- `[10,11]` = `0.25 * (0.25 + 0.75)` = `0.25`
- `[10,10]` = `0.25 * (0.25 + 0.25)` = `0.125`

That makes sense to me? The voxels sum to 1, and it's more heavily weighted towards `[11,11]`.

Reply:
Actually, the more I think about it, the more I'm wondering whether the weighting is correct? 

For example, say it was `[10.9, 10.9]` instead:

- `[11,11]` = `0.25 * (0.90 + 0.10)` = `0.45`
- `[11,10]` = `0.25 * (0.90 + 0.10)` = `0.25`
- `[10,11]` = `0.25 * (0.10 + 0.90)` = `0.25`
- `[10,10]` = `0.25 * (0.10 + 0.10)` = `0.05`

Or even say it was `[10.99, 10.99]` instead:

- `[11,11]` = `0.25 * (0.99 + 0.99)` = `0.495`
- `[11,10]` = `0.25 * (0.99 + 0.01)` = `0.25`
- `[10,11]` = `0.25 * (0.01 + 0.99)` = `0.25`
- `[10,10]` = `0.25 * (0.01 + 0.01)` = `0.005`

Maybe my math skills are lacking here, but this doesn't make intuitive sense to me?

Since `[10.99,10.99]` is practically `[11,11]`, shouldn't most of the weighting be put on the `[11,11]` voxel? Instead, there's an upper bound of `0.5` for each voxel, and the corner voxels will always add to `0.25`, which doesn't seem right  to me... :thinking:

Reply:
Hey @joshuacwnewton, thanks for looking into the math in this much detail! We actually considering the 0.25 weighting for pixel too but, as you also ended up finding through your calculations, if the actual centerline coordinates are `[10.9, 10.9]` then we would assume that the pixel `[11, 11]` will have the highest intensity. We are somewhat seeing that when we choose 0.5 as the weighting. With 0.25, the intensities are just too small for a pixel and it didn't make sense to us as well!

@NathanMolinier had suggested an alternative where the diagonals sum up to 1, that is, if the `[11, 11]` has the highest intensity, then the diagonally opposite pixel (usually the one with (floor, floor)) should have the least intensity. This is what we implemented here.

Reply:
Moreover, from the screenshots that @valosekj posted at https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4026#issue-1568820188, looking at the intensity of the binarized centerline, we were thinking that our soft centerline should have similar intensity at that pixel if not as high as the binarized. Hence, if we choose 0.25 as the weighting, the 2 pixels (left and the top) will have similar intensities, and we wouldn't be sure where the actual centerline is. 

(Also, note that I'm just trying to explain our thought process as we did the math on a paper. The intention is not to sound too defensive! Let me know if any of what I said doesn't make sense! Thanks again for looking into the details, this is really useful!

Reply:
I have two thoughts here:

#### 1.

I don't think my concern from https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4026#discussion_r1095891109 has been addressed?

My understanding of partial volume calculations was that the "soft" segmentation must sum to the same values as the "binarized" segmentation (in this case, 1). But, right now the 2x2 grid sums to 2 instead, which breaks the requirement for partial volume calculations. (I think?)

#### 2.

> Hence, if we choose 0.25 as the weighting, the 2 pixels (left and the top) will have similar intensities, and we wouldn't be sure where the actual centerline is.

You're right, this is a problem. However, I think that the problem you describe is independent of the constant scaling factor; whether we use `0.25` or `0.5` shouldn't matter, because the scaling factor is applied equally to all 4 voxel locations. It's something that can be factored out?

To fix the problem of "unclear weighting", I think the actual "distance" computation (i.e. x_diff, y_diff) needs to be updated, rather than the constant scaling factor.

Reply:
I believe the following formulas would be better, because:
- the weights would still add up to 1 (up to a tiny floating point error); and
- the initial floating point coordinates of the centerline will be the weighted average of the rounded coordinates (with the given weights/probabilities). (In other words, coordinates like `[10.9, 10.9]` and `[11.1, 11.1]` would put almost all the weight on coordinate `[11, 11]`.)

Additionally:
- In both cases (`param.soft == 0` and `param.soft == 1`), it would be wise to clip the centerline coordinates to the ranges `[0, xdim-1]` and `[0, ydim-1]`, because other values (including `xdim` and `ydim` themselves) would cause array indexing errors.
- Note that the expected relationship `ceil(x) == floor(x)+1` fails when `x` happens to be an integer.
  - This is actually good for us, because using just `floor(x)` and `floor(x)+1` would run the risk of getting outside the bounds `[0, xdim-1]`.
  - But we have to be careful, otherwise we might _overwrite_ one of the probabilities with another one, and the total would become less than 1. The way to fix this is to add, rather than assign, the probabilities.

```py3
x_max, y_max = im_centerline.data.shape[:2]
x_centerline_fit = x_centerline_fit.clip(0, x_max-1)
y_centerline_fit = y_centerline_fit.clip(0, y_max-1)

x_floor = np.floor(x_centerline_fit).astype(int)
y_floor = np.floor(y_centerline_fit).astype(int)

x_ceil = np.ceil(x_centerline_fit).astype(int)
y_ceil = np.ceil(y_centerline_fit).astype(int)

x_frac = x_centerline_fit - x_floor
y_frac = y_centerline_fit - y_floor

im_centerline.data[x_floor, y_floor, z_ref] += (1-x_frac) * (1-y_frac)
im_centerline.data[x_floor, y_ceil, z_ref] += (1-x_frac) * y_frac
im_centerline.data[x_ceil, y_floor, z_ref] += x_frac * (1-y_frac)
im_centerline.data[x_ceil, y_ceil, z_ref] += x_frac * y_frac
```

Reply:
So, in this case, for the coordinate `[10.99,10.99]`, we get:

<details>

* `floor`: 10
* `ceil`: 11
* `frac`: 0.99
* `[10,10]`: 0.01 * 0.01 = 0.0001
* `[10,11]`: 0.01 * 0.99 = 0.0099
* `[11,10]`: 0.99 * 0.01 = 0.0099
* `[11,11]`: 0.99 * 0.99 = 0.9801

And, 0.0001 + 0.0099 + 0.0099 + 0.9801 = 1!

</details>

This matches my expectations exactly! :tada:

Reply:
Thank you @mguaypaq for your suggestion! I think this covers the issues we have been having! 

There is one question though - Do you think the `x_max - 1` in `x_centerline_fit = x_centerline_fit.clip(0, x_max-1)` is necessary? Because, in the original implementation, we have this `round_and_clip(x_centerline_fit, clip=[0, im_centerline.data.shape[0]])` and it is clipping until what would be `x_max`.

Reply:
> There is one question though - Do you think the `x_max - 1` in `x_centerline_fit = x_centerline_fit.clip(0, x_max-1)` is necessary? Because, in the [original implementation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/0f228da127cc78408462b824624564ea7e09a7ba/spinalcordtoolbox/centerline/core.py#L181-L182), we have this `round_and_clip(x_centerline_fit, clip=[0, im_centerline.data.shape[0]])` and it is clipping until what would be `x_max`.

Yes, I think the `-1` is necessary, and I think that's a (very minor) bug in the original. For example:
```py3
>>> import numpy as np
>>>
>>> # Make a 2x4 numpy array
>>> data = np.arange(8).reshape((2, 4)); data
array([[0, 1, 2, 3],
       [4, 5, 6, 7]])
>>>
>>> # Get coordinates from its shape
>>> x_max, y_max = data.shape; data.shape
(2, 4)
>>>
>>> # We can access the last element with -1's
>>> data[x_max-1, y_max-1]
7
>>>
>>> # But without the -1's we get an IndexError
>>> data[x_max, y_max]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: index 2 is out of bounds for axis 0 with size 2
```

Reply:
Thanks all for the fruitful discussion! I implemented [Mathieu's solution](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4026#discussion_r1096093021) in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4026/commits/fcdcef65c0d2b0cef47235e88b5b83e699653079. 


> Yes, I think the `-1` is necessary, and I think that's a (very minor) bug in the original. For example:

Just for clarification. What do you mean by "the original"? Because [round_and_clip](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/0f228da127cc78408462b824624564ea7e09a7ba/spinalcordtoolbox/centerline/core.py#L276) function `-1` contains:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/0f228da127cc78408462b824624564ea7e09a7ba/spinalcordtoolbox/centerline/core.py#L276

Reply:
> Because [round_and_clip](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/0f228da127cc78408462b824624564ea7e09a7ba/spinalcordtoolbox/centerline/core.py#L276) function `-1` contains:

I stand corrected! I just read the help strings for `np.round` and `np.clip` and assumed that a function called `round_and_clip` operated the same way, but I should have read the code for `round_and_clip` as well.

Reply:
Aside: The existence of the `round_and_clip` function is confusing to me. What is its purpose? To trip up developers? :stuck_out_tongue_winking_eye: 

It's literally just a mashup of the `round` and `clip` functions -- it will do one or the other, but not both, so there's not really any added benefit/functionality. 

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/0f228da127cc78408462b824624564ea7e09a7ba/spinalcordtoolbox/centerline/core.py#L275-L278

Why not just call `np.round` and `np.clip` individually when needed?

Plus, `round_and_clip` is only truly called in the context of `centerline/core.py`, so its usage isn't even that widespread:

![image](https://user-images.githubusercontent.com/16181459/217347842-e9beab93-5546-40b9-ad5b-63443c21f901.png)

Given that it's a source of confusion, I would suggest to get rid of it entirely?